crates/bevy_ui/src/lib.rs

```diff
@@ -75,6 +78,100 @@ impl Default for UiScale {
     }
 }
 
+/// General representation of progress between two values.
+#[derive(Debug)]
+pub struct Progress<T>
+where
+    T: Send + Sync + Copy,
+{
+    /// The minimum value that the progress can have, inclusive.
+    min: T,
+    /// The maximum value that the progress can have, inlucsive.
+    max: T,
+    /// The current value of progress.
+    value: T,
+}
+
+impl<T: Send + Sync + Copy> Progress<T>
+where
+    T: PartialOrd<T>,
+{
+    /// Creates a new progress using a value, min and max.
+    ///
+    /// The value must be within min and max.
+    /// Will panic if value is out-of-bounds.
+    pub fn new(value: T, min: T, max: T) -> Self {
+        Self::from_range(value, min..=max)
+    }
+
+    /// Creates a new progress using a value and a range.
+    ///
+    /// The value must be within the bounds of the range.
+    /// Will panic if value is out-of-bounds.
+    pub fn from_range(value: T, range: RangeInclusive<T>) -> Self {
+        if range.contains(&value) {
+            Self {
+                value,
+                min: *range.start(),
+                max: *range.end(),
+            }
+        } else {
+            panic!("Tried creating progress with value out of bounds.");
+        }
+    }
+
+    /// Gets the min bound of the progress.
+    pub fn min(&self) -> T {
+        self.min
+    }
+
+    /// Gets the max bound of the progress.
+    pub fn max(&self) -> T {
+        self.max
+    }
+
+    /// Gets the bounds of the progress.
+    pub fn bounds(&self) -> RangeInclusive<T> {
+        self.min..=self.max
+    }
+
+    /// Gets the current value of progress.
+    pub fn progress(&self) -> T {
+        self.value
+    }
+
+    /// Sets the progress to a new value and returns the new value if successful.
+    ///
+    /// Value must be between `min` and `max`.
+    pub fn set_progress(&mut self, new_value: T) -> Result<T, &str> {
+        if self.bounds().contains(&new_value) {
+            self.value = new_value;
+            Ok(self.value)
+        } else {
+            Err("Value outside the bounds of the Progress")
```

Comment:
I would consider making a `ProgressError` enum for errors connected to the `Progress` struct, it's just my preference though, I will still approve it either way.

Reply:
Totally agree here.